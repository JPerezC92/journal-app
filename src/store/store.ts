import { configureStore } from "@reduxjs/toolkit";
import { uiReducer, authReducer, notesReducer } from "../reducers";

export const reducers = {
  authReducer,
  uiReducer,
  notesReducer,
};

export const store = configureStore({
  reducer: reducers,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
