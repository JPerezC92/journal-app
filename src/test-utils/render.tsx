import { configureStore } from "@reduxjs/toolkit";
import { render, RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { uiReducer, authReducer, notesReducer } from "../reducers";

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: RenderOptions<
    typeof import("/home/philip/projects/journal-app/node_modules/@testing-library/dom/types/queries"),
    HTMLElement
  >
) => {
  const store = configureStore({
    reducer: {
      authReducer,
      uiReducer,
      notesReducer,
    },
  });
  const reduxStore: React.FC = ({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  return { component: render(ui, { wrapper: reduxStore, ...options }), store };
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
