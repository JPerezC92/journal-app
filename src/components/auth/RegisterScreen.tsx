import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks";
import { uiActions } from "../../reducers";
import { startRegister } from "../../reducers/authReducer";
import { ValidatorService } from "../../services";

import { RootState } from "../../store/store";

const RegisterScreen = () => {
  const validatorService = new ValidatorService();
  const { errorMessage } = useSelector((state: RootState) => state.uiReducer);
  const ui = useSelector((state: RootState) => state.uiReducer);

  const dispatch = useDispatch();

  const { formValues, handleInputChange } = useForm({
    name: "",
    email: "",
    password: "",
    confirmPassword: "",
  });
  const { name, email, password, confirmPassword } = formValues;

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { success, errorMessage } =
      validatorService.validateRegisterForm(formValues);

    if (success) {
      dispatch(startRegister(formValues));
    } else {
      dispatch(uiActions.setError(errorMessage));
    }
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>

      {errorMessage && <div className="auth__alert-error">{errorMessage}</div>}

      <form
        aria-label="form"
        onSubmit={handleSubmit}
        className="animate__animated animate__fadeIn animate__faster"
      >
        <input
          className="auth__input"
          type="name"
          placeholder="Name"
          name="name"
          autoComplete="off"
          onChange={handleInputChange}
          value={name}
        />

        <input
          className="auth__input"
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          onChange={handleInputChange}
          value={email}
        />

        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          onChange={handleInputChange}
          value={password}
        />

        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="confirmPassword"
          onChange={handleInputChange}
          value={confirmPassword}
        />

        <button
          className="btn btn-primary btn-block mb-5"
          type="submit"
          disabled={ui.loading}
        >
          Register
        </button>

        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
