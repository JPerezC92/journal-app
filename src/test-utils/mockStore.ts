import {
  AnyAction,
  CombinedState,
  combineReducers,
  configureStore,
  createSlice,
} from "@reduxjs/toolkit";
import { reducers } from "../store/store";

const storeCleanSlice = createSlice({
  name: "store",
  initialState: {},
  reducers: {
    clean: () => {},
  },
});

export const storeCleanActions = storeCleanSlice.actions;
const storeCleanReducer = storeCleanSlice.reducer;
const combinedReducers = combineReducers({ ...reducers, storeCleanReducer });

const testReducers = (
  state: CombinedState<any> | undefined,
  action: AnyAction
) => {
  if (action.type === "store/clean") {
    state = undefined;
  }
  return combinedReducers(state, action);
};

export const mockStore = configureStore({
  reducer: testReducers,
});

export type MockRootState = ReturnType<typeof mockStore.getState>;
export type MockAppDispatch = typeof mockStore.dispatch;
