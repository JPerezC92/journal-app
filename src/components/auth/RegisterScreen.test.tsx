import renderer, { act } from "react-test-renderer";
import { fireEvent, RenderResult } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { store } from "../../store/store";
import { render } from "../../test-utils";
import RegisterScreen from "./RegisterScreen";

function insertFormValues(
  component: RenderResult<
    typeof import("/home/philip/projects/journal-app/node_modules/@testing-library/dom/types/queries"),
    HTMLElement
  >
) {
  const nameField = component.getByPlaceholderText("Name");
  fireEvent.change(nameField, { target: { value: "testing new user" } });

  const emailField = component.getByPlaceholderText("Email");
  fireEvent.change(emailField, {
    target: { value: "testing_new_user@gmail.com" },
  });

  const passwordField = component.getByPlaceholderText("Password");
  fireEvent.change(passwordField, { target: { value: "123456" } });

  const confirmPasswordField =
    component.getByPlaceholderText("Confirm password");
  fireEvent.change(confirmPasswordField, { target: { value: "123456" } });
}

describe("Test on <RegisterScreen />", () => {
  test("should match the snapshot correctly", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <RegisterScreen />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("should register a new user", () => {
    const { component, store } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    jest.spyOn(store, "dispatch");

    insertFormValues(component);

    const button = component.getByRole("button");
    act(() => {
      fireEvent.click(button);
    });

    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  test("should display an email error", () => {
    const { component } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    insertFormValues(component);

    const emailField = component.getByPlaceholderText("Email");
    fireEvent.change(emailField, { target: { value: "" } });

    const button = component.getByRole("button");

    act(() => {
      fireEvent.click(button);
    });
    const errorDiv = component.getByText("Email invalid");

    expect(errorDiv).toHaveTextContent("Email invalid");
  });

  test("should display an name error", () => {
    const { component } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    insertFormValues(component);

    const nameField = component.getByPlaceholderText("Name");
    fireEvent.change(nameField, { target: { value: "" } });

    const button = component.getByRole("button");
    act(() => {
      fireEvent.click(button);
    });

    const errorDiv = component.getByText("Name is required");
    expect(errorDiv).toHaveTextContent("Name is required");
  });

  test("should display a password error", () => {
    const { component } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    insertFormValues(component);

    const passwordField = component.getByPlaceholderText("Password");
    const button = component.getByRole("button");
    fireEvent.change(passwordField, { target: { value: "" } });

    act(() => {
      fireEvent.click(button);
    });

    const errorDiv = component.getByText(
      "Password should be at least 5 characters"
    );

    expect(errorDiv).toHaveTextContent(
      "Password should be at least 5 characters"
    );
  });

  test("should display a confirmPassword error", () => {
    const { component } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    insertFormValues(component);

    const confirmPasswordField =
      component.getByPlaceholderText("Confirm password");
    fireEvent.change(confirmPasswordField, { target: { value: "" } });

    const button = component.getByRole("button");
    act(() => {
      fireEvent.click(button);
    });

    const errorDiv = component.getByText("Password should match each other");

    expect(errorDiv).toHaveTextContent("Password should match each other");
  });

  test("should display a passwordd error", () => {
    const { component } = render(
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    );

    insertFormValues(component);

    const passwordField = component.getByPlaceholderText("Password");
    fireEvent.change(passwordField, { target: { value: "aasddas" } });

    const button = component.getByRole("button");
    act(() => {
      fireEvent.click(button);
    });

    const errorDiv = component.getByText("Password should match each other");

    expect(errorDiv).toHaveTextContent("Password should match each other");
  });
});
