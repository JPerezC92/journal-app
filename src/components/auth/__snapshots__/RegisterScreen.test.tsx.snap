// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <RegisterScreen /> should match the snapshot correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    aria-label="form"
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="name"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
